include/fi_perf.h

```diff
@@ -0,0 +1,386 @@
+ * distributions including binary code include the above copyright notice and
+ * this paragraph in its entirety in the documentation or other materials
+ * provided with the distribution
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+/* RDPMC specific logic */
+
+#include <unistd.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <sys/fcntl.h>
+#include <linux/perf_event.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <asm/unistd.h>
+
+struct fi_perf_rdpmc_context {
+	int fd;
+	struct perf_event_mmap_page *buf;
+	uint32_t type;
+	uint32_t config;
+};
+
+#if defined(__ICC) || defined(__INTEL_COMPILER)
+#include "immintrin.h"
+#endif
+
+#define fi_perf_rdpmc_rmb() asm volatile("" ::: "memory")
+
+/* Default configuration */
+
+#define FI_PERF_RDPMC_DEFAULT_TYPE   (PERF_TYPE_HARDWARE)
+#define FI_PERF_RDPMC_DEFAULT_CONFIG (PERF_COUNT_HW_CPU_CYCLES)
+#define FI_PERF_RDPMC_ALIGNMENT      (4096)
+
+static inline int fi_perf_init(struct fi_perf_handle ** handle_ptr)
+{
+	struct fi_perf_handle * handle = *handle_ptr;
+
+	int result = 0;
+
+	result = posix_memalign((void **)&handle, FI_PERF_RDPMC_ALIGNMENT,
+							sizeof(struct fi_perf_handle));
+	if (result != 0) {
+		/* Error handling */
+		goto exit_err_0;
+	}
+
+	memset(handle, 0, sizeof(struct fi_perf_handle));
+	handle->count = FI_PERF_SLOT_COUNT_MAX;
+
+	struct fi_perf_rdpmc_context * context =
+		(struct fi_perf_rdpmc_context *)handle->private;
+
+	int env_result    = 1;
+	char * env_type   = NULL;
+	char * env_config = NULL;
+
+	env_type = getenv("FI_PERF_RDPMC_TYPE");
+	if (env_type) {
+		context->type = (uint32_t)strtoll(env_type, NULL, 16);
+	} else {
+		env_result = 0;
+	}
+
+	env_config = getenv("FI_PERF_RDPMC_CONFIG");
+	if (env_config) {
+		context->config = (uint32_t)strtoll(env_config, NULL, 16);
+	} else {
+		env_result = 0;
+	}
+
+	if (!env_result) {
+		context->type   = FI_PERF_RDPMC_DEFAULT_TYPE;
+		context->config = FI_PERF_RDPMC_DEFAULT_CONFIG;
+	}
+
+	struct perf_event_attr attr = {
+		.type = context->type,
+		.size = sizeof(struct perf_event_attr),
+		.config = context->config,
+		.sample_type = PERF_SAMPLE_READ,
+	};
+
+	context->fd = syscall(__NR_perf_event_open, &attr, 0, -1, -1, 0);
+	if (context->fd < 0) {
+		perror("perf_event_open");
+		/* Error handling */
+		goto exit_err_1;
+	}
+
+	context->buf = mmap(NULL, sysconf(_SC_PAGESIZE),
+						PROT_READ, MAP_SHARED, context->fd, 0);
+	if (context->buf == MAP_FAILED) {
+		perror("mmap on perf fd");
+		/* Error handling */
+		goto exit_err_2;
+	}
+
+ exit:
+	*handle_ptr = handle;
+	return result;
+
+	/* Error handling */
+ exit_err_2:
+	close(context->fd);
+ exit_err_1:
+	free(handle);
+ exit_err_0:
+	handle = NULL;
+
+	result = -1;
+	goto exit;
+}
+
+static inline void fi_perf_finalize(struct fi_perf_handle * handle)
+{
+	if (handle) {
```

Comment:
nit: I prefer to just assert(handle), rather than have the entire function indented under an if() check.

Reply:
fixed